### PR TITLE
Widget & unit testing

### PR DESCRIPTION
**Pull request by Ramakrishna530**
- Added Widget Tests
- Added Unit tests

### CODE REVIEW

File: lib/provider/repositories_provider.dart [hunk 1]
```diff
@@ -4,20 +4,14 @@ import '../core/http/api_response.dart';
 import '../models/get_repositories/repository_response.dart';
 import '../repository/get_repositories_repo.dart';
 
-abstract class RepositoriesProvider {
```
Comment (gabrielgarciagava) [inline lib/provider/repositories_provider.dart:7]: One thing I would like yoy to try.
- Add this abstract class back
- Make it implements Listenable.
- You may need to implement some other methods on RepositoriesProviderImpl.
- Where you are providing via ChangeNotifierProvider, change to ListenableProvider<RepositoriesProvider>

This way I think we can get back the interface, and use only it everywhere.

File: test/provider/repositories_provider_test.dart [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    late MockGetRepositoriesRepo mockGetRepositoriesRepo;
+    const language = "Dart";
+    setUp(() {
+      mockGetRepositoriesRepo = MockGetRepositoriesRepo();
+      repositoriesProviderImpl = RepositoriesProviderImpl(
+          getRepositoriesRepo: mockGetRepositoriesRepo);
+    });
+
+    List<RepositoryResponse> getMockRepositories() {
+      final file = File('test_resources/repositories.json');
```
Comment (gabrielgarciagava) [inline test/provider/repositories_provider_test.dart]: For this layer, I don't think you should depend on how the json looks like. The conversion of json to `List<RepositoryResponse>` is already handled and test in `GetRepositoriesRepo`.
So here, I would recommend you to build `List<RepositoryResponse>` directly.
```
[
  RepositoryResponse(xxx), // or RepositoryResponse.testInstance()
]

File: test/view/screens/repositories_screen_test.dart [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+@GenerateMocks([GetRepositoriesRepo])
+void main() {
+  group("$RepositoriesScreen", () {
+    late RepositoriesProviderImpl repositoriesProvider;
+    late MockGetRepositoriesRepo mockGetRepositoriesRepo;
+
+    setUp(() {
+      mockGetRepositoriesRepo = MockGetRepositoriesRepo();
+      repositoriesProvider = RepositoriesProviderImpl(
+          getRepositoriesRepo: mockGetRepositoriesRepo);
```
Comment (gabrielgarciagava) [inline test/view/screens/repositories_screen_test.dart]: missing ,

File: test/view/screens/repositories_screen_test.dart [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        final repositoryOwnerResponse =
+            RepositoryOwnerResponse(avatarUrl: "avatarUrl-$i");
+        final repositoryResponse = RepositoryResponse(
+            id: i,
+            name: "name-$i",
+            fullName: "fullName-$i",
+            private: false,
+            owner: repositoryOwnerResponse,
+            watchersCount: i,
+            score: i.toDouble());
```
Comment (gabrielgarciagava) [inline test/view/screens/repositories_screen_test.dart]: missing ,

File: test/view/screens/repositories_screen_test.dart [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      when(mockGetRepositoriesRepo.getRepositories(language: "Dart"))
+          .thenAnswer((_) => Future.value(mockRepositoriesResponse));
+      await tester.pumpWidget(createMaterialApp());
+      expect(find.text("Loading..."), findsOneWidget);
+      await tester.pumpAndSettle(const Duration(milliseconds: 1));
+      for (final repositoryResponse in mockRepositoriesResponse) {
+        expect(find.text(repositoryResponse.name), findsOneWidget);
+        expect(
+            find.text("Watchers Count - ${repositoryResponse.watchersCount}"),
+            findsOneWidget);
```
Comment (gabrielgarciagava) [inline test/view/screens/repositories_screen_test.dart]: lot of missing , :P

File: test/view/screens/repositories_screen_test.dart [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import 'repositories_screen_test.mocks.dart';
+
+@GenerateMocks([GetRepositoriesRepo])
+void main() {
+  group("$RepositoriesScreen", () {
+    late RepositoriesProviderImpl repositoriesProvider;
+    late MockGetRepositoriesRepo mockGetRepositoriesRepo;
+
+    setUp(() {
+      mockGetRepositoriesRepo = MockGetRepositoriesRepo();
```
Comment (gabrielgarciagava) [inline test/view/screens/repositories_screen_test.dart]: Here you are mocking the wrong layer

Now:
Screen -> Provider -> Repo (mocked)

Expected:
Screen -> Provider (mocked)

File: test/view/screens/repositories_screen_test.dart [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      await tester.tap(dropdown);
+      await tester.pump();
+      final menuItem = find.text(language).last;
+      await tester.tap(menuItem);
+      await tester.pumpAndSettle(const Duration(milliseconds: 1));
+      for (final repositoryResponse in mockRepositoriesResponse) {
+        expect(find.text(repositoryResponse.name), findsOneWidget);
+        expect(
+            find.text("Watchers Count - ${repositoryResponse.watchersCount}"),
+            findsOneWidget);
```
Comment (gabrielgarciagava) [inline test/view/screens/repositories_screen_test.dart]: Increase your IDE max columns to 120, this is what we use in our project.

File: test/view/widget/drop_down_test.dart [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      "JavaScript",
+      "Kotlin",
+      "PHP",
+      "Python",
+      "Ruby",
+      "Swift",
+      "TypeScript",
+    ];
+
+    String? selectedItem;
```
Comment (gabrielgarciagava) [inline test/view/widget/drop_down_test.dart]: Not so nice to have this kind of global state.

If you have two tests testing didSelect, you would need to reset this guy on a tearDown, for example. So one test does not influence the other.

Move it to the test that is using it.
On the other tests where you just need to inject the function, inject a very simple lambda that does not do anything.

File: test/provider/repositories_provider_test.dart (hunk 2 above)
Comment (Ramakrishna530) [reply test/provider/repositories_provider_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 3 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 4 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 7 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 5 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 7 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: test/view/widget/drop_down_test.dart (hunk 8 above)
Comment (Ramakrishna530) [reply test/view/widget/drop_down_test.dart]: done

File: test/view/screens/repositories_screen_test.dart (hunk 5 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

File: lib/provider/repositories_provider.dart (hunk 1 above)
Comment (Ramakrishna530) [reply lib/provider/repositories_provider.dart:7]: done. Thanks for your help on this.

File: test/view/screens/repositories_screen_test.dart (hunk 6 above)
Comment (Ramakrishna530) [reply test/view/screens/repositories_screen_test.dart]: done

**Review by gabrielgarciagava — APPROVED** [review]

**Pull request merged**